packages/tre/src/storage2/keyvalue.ts

```diff
@@ -0,0 +1,255 @@
+}
+
+export interface KeyEntriesQuery {
+  prefix?: string;
+  cursor?: string;
+  limit: number;
+}
+export interface KeyEntries<Metadata = unknown> {
+  keys: KeyEntry<Metadata>[];
+  cursor?: string;
+}
+
+interface Row {
+  key: string;
+  blob_id: string;
+  expiration: number | null; // milliseconds since unix epoch
+  metadata: string | null;
+}
+const SQL_SCHEMA = `
+CREATE TABLE IF NOT EXISTS _mf_entries (
+  key TEXT PRIMARY KEY,
+  blob_id TEXT NOT NULL,
+  expiration INTEGER,
+  metadata TEXT
+);
+CREATE INDEX IF NOT EXISTS _mf_entries_expiration_idx ON _mf_entries(expiration);
+`;
+function sqlStmts(db: TypedDatabase) {
+  const stmtGetBlobIdByKey = db.prepare<Pick<Row, "key">, Pick<Row, "blob_id">>(
+    "SELECT blob_id FROM _mf_entries WHERE key = :key"
+  );
+  const stmtPut = db.prepare<Row>(
+    `INSERT OR REPLACE INTO _mf_entries (key, blob_id, expiration, metadata)
+    VALUES (:key, :blob_id, :expiration, :metadata)`
+  );
+
+  return {
+    getByKey: db.prepare<Pick<Row, "key">, Row>(
+      "SELECT key, blob_id, expiration, metadata FROM _mf_entries WHERE key = :key"
+    ),
+    put: db.transaction((newEntry: Row) => {
+      // TODO(perf): would be really nice if we didn't have to use a transaction
+      //  here, need old blob ID for cleanup
+      const key = newEntry.key;
+      const entry = stmtGetBlobIdByKey.get({ key });
+      stmtPut.run(newEntry);
+      return entry?.blob_id;
+    }),
+    deleteByKey: db.prepare<
+      Pick<Row, "key">,
+      Pick<Row, "blob_id" | "expiration">
+    >("DELETE FROM _mf_entries WHERE key = :key RETURNING blob_id, expiration"),
+    deleteExpired: db.prepare<{ now: number }, Pick<Row, "blob_id">>(
+      // `expiration` may be `NULL`, but `NULL < ...` should be falsy
+      "DELETE FROM _mf_entries WHERE expiration < :now RETURNING blob_id"
+    ),
+    list: db.prepare<
+      {
+        now: number;
+        escaped_prefix: string;
+        start_after: string;
+        limit: number;
+      },
+      Omit<Row, "blob_id">
+    >(
+      `SELECT key, expiration, metadata FROM _mf_entries
+        WHERE key LIKE :escaped_prefix || '%' ESCAPE '\\'
+        AND key > :start_after
+        AND (expiration IS NULL OR expiration >= :now)
+        ORDER BY key LIMIT :limit`
+    ),
+  };
+}
+
+function escapePrefix(prefix: string) {
+  // Prefix all instances of `\`, `_` and `%` with `\`
+  return prefix.replaceAll(/[\\_%]/g, "\\$&");
+}
+
+function rowEntry<Metadata>(entry: Omit<Row, "blob_id">): KeyEntry<Metadata> {
+  return {
+    key: entry.key,
+    expiration: entry.expiration ?? undefined,
+    metadata: entry.metadata === null ? undefined : JSON.parse(entry.metadata),
+  };
+}
+
+export class KeyValueStorage<Metadata = unknown> {
+  readonly #stmts: ReturnType<typeof sqlStmts>;
+
+  constructor(
+    private readonly storage: NewStorage,
+    private readonly clock = defaultClock
+  ) {
+    storage.db.pragma("case_sensitive_like = TRUE");
```

Comment:
Should this be defined with the construction of the database itself? Where the WAL is set?